pkg/hostpath/v1alpha1/hostpath.go

```diff
@@ -0,0 +1,157 @@
+	"path/filepath"
+	"strings"
+)
+
+// HostPath is a wrapper over a given path string
+// It provides build, validations and other common
+// logic to be used by various feature specific callers.
+type HostPath string
+
+// Predicate defines an abstraction
+// to determine conditional checks
+// against the provided HostPath instance
+type Predicate func(HostPath) bool
+
+// IsValidPath is a predicate that determines
+// the hostpath is a well formed path
+func IsValidPath() Predicate {
+	return func(hp HostPath) bool {
+		//Validate that path is well formed.
+		//_, err := filepath.Abs(string(hp))
+		//if err != nil {
+		//	return false
+		//}
+		//return true
+		return filepath.IsAbs(string(hp))
+	}
+}
+
+// IsNonRoot is a predicate that determines
+// the hostpath is not directly under /
+func IsNonRoot() Predicate {
+	return func(hp HostPath) bool {
+		//Validate that hostpath is not directly under "/"
+		path := strings.TrimSuffix(string(hp), "/")
+		parentDir, subDir := filepath.Split(path)
+		parentDir = strings.TrimSuffix(parentDir, "/")
+		subDir = strings.TrimSuffix(subDir, "/")
+		if parentDir == "" || subDir == "" {
+			// it covers the `/` case
+			return false
+		}
+		return true
+	}
+}
+
+// Builder provides utility functions
+// on the HostPath to extract different information
+type Builder struct {
+	path   HostPath
+	checks map[*Predicate]string
+}
+
+// NewBuilder builds a new Builder
+func NewBuilder() *Builder {
+	return &Builder{
+		path:   "",
+		checks: make(map[*Predicate]string),
+	}
+}
+
+// WithPath initializes the Builder with the path on
+// which it has to operate
+func (b *Builder) WithPath(path string) *Builder {
+	b.path = HostPath(path)
+	return b
+}
+
+// WithPathJoin initializes the Builder with the path
+// by joining the arguments
+func (b *Builder) WithPathJoin(basePath, relPath string) *Builder {
+	b.path = HostPath(filepath.Join(basePath, relPath))
+	return b
+}
+
+// WithChecks adds the list of Predicates to
+// the list of checks to be performed by the builder.
+func (b *Builder) WithChecks(checks ...Predicate) *Builder {
+	for _, check := range checks {
+		b.WithCheck(check)
+	}
+	return b
+}
+
+// WithCheck add a single Predicate to the
+// list of checks to be performed by the builder
+func (b *Builder) WithCheck(check Predicate) *Builder {
+	return b.WithCheckf(check, "")
+}
+
+// WithCheckf adds a single Predicate to the
+// list of checks to be performed by the builder, along
+// with the message to be returned when the check fails
+func (b *Builder) WithCheckf(check Predicate, msg string, args ...interface{}) *Builder {
+	b.checks[&check] = fmt.Sprintf(msg, args...)
+	return b
+}
+
+// Validate runs through all the checks set on the
+// builder and returns an error when any of the checks fails.
+func (b *Builder) Validate() error {
+	if b.path == "" {
+		return errors.New("failed to validate: missing host path")
+	}
+
+	b = b.WithCheckf(IsValidPath(), "invalid path %v", b.path)
```

Comment:
There might be a chance where caller adds `IsValidPath()` check via this builder.
It might result in duplicate errors.

Reply:
make IsValidPath private.